### DISCUSSION TITLE
Chat-Sample context access

### DISCUSSION THREAD

**Discussion by jzworkman** [microsoft/vscode-discussions#1114] [Extension Development QnA] [answered]
Right now when using the chat-sample and proposed chatParticipant, chatVariableResolver, and languageModels apis, when using the ChatParticipant there is only access to variables(if the user for example uses #selection). 

However, when using chat normally, it selects the context based on visible window/selected code automatically.  This is leading to different outcomes when using the chat participant and not.  

While we can use something like the below to do variable resolution.  This requires the user to always remember to use the #selection variable in their prompts, which is something they don't need to do for non-participant flows

```
const selection = request.variables.find(x => x.name === 'selection');
var prompt = request.prompt;
if (selection && selection.range && selection.values.length > 0) {
       prompt = prompt.substring(0, selection.range[0]) + selection.values[0].value + prompt.substring(selection.range[1]);
}
```

Is there an example or path to be able to utilize the same context selection that Copilot and Copilot chat extensions use?  Otherwise I feel the ChatParticipant solution at present is preventing adding the system context, since the user prompt that gets passed along is missing the reference material it would normally have

@roblourens I know you have done most the work in the chat-sample example.  Have you seen these same limitations and is there a roadmap for future API changes for this?

**Comment by roblourens** [comment, marked as answer]
For now, you can only read context yourself using all the methods of the extension API. This is what the Copilot Chat extension does.

In the future, you may be able to pull in a variable manually from your participant, but don't count on that soon.
